proposals/0000-specialise-pragma-with-type-applications.rst

```diff
@@ -0,0 +1,127 @@
+                 implements the described feature.
+
+.. highlight:: haskell
+
+Proposal title
+==============
+
+GHC's ``SPECIALISE`` pragma allows the user to instruct the compiler specialise
+a polymorphic function at a concrete type. However, its current syntax is quite
+unwieldy, requiring that the user repeat the entire type of the the specialised
+function, including those parts that are shared with its unspecialised type.
+
+Here we exploit the ``TypeApplications`` syntax added in GHC 8.0 to write many
+uses of ``SPECIALISE`` more concisely.
+
+Motivation
+----------
+
+The ``SPECIALISE`` pragma is a means of guiding the optimizer towards the types
+a function will most commonly be used at. For instance, consider this
+polymorphic function from the ``postgresql-simple`` database bindings, ::
+
+    foldWithParser :: ( MonadIO m, MonadMask m, ToRow params )
+                   => RowParser row -> Connection
+                   -> Query -> params
+                   -> a -> (a -> row -> m a)
+                   -> m a
+    foldWithParser parser conn template qs a f = ...
+
+Here we have a function with a non-trivial, polymorphic type. If we know that
+this function may often be used at the ``IO`` monad then we might instruct
+GHC to derive an appropriate specialisation, ::
+
+    {-# SPECIALISE
+        foldsAndParser
+            :: RowParser row -> Connection
+            -> Query -> params
+            -> a
+            -> (a -> row -> IO a) -> IO a
+        #-}
+
+This is, of course, quite a mouthful! Moreover, if we would at some point like
+to change the type of ``foldWithParser``, we would need to updae each of the
+specialisations.
+
+We would ideally like to use the fact that we are merely instantiating the
+function's ``m`` type argument at ``IO``.
+
+Proposed Change
+---------------
+
+Here we propose to allow use of the ``TypeApplications`` syntax introduced in
+GHC 8.0.1 in place of the type in a ``SPECIALISE`` pragma. For instance, in the
+case of ``foldWithParser`` above we might first add an explicit ``forall`` to
+make the order of type variables clear, ::
+
+
+    foldWithParser :: forall m params row a. ( MonadIO m, MonadMask m, ToRow params )
+                   => RowParser row -> Connection
+                   -> Query -> params
+                   -> a -> (a -> row -> m a)
+                   -> m a
+    foldWithParser parser conn template qs a f = ...
+
+and then write its ``SPECIALISE`` pragma as a unsaturated type application, ::
+
+    {-# SPECIALISE foldsAndParser @IO #-}
+
+Of course, this raises the question of how we might quantify over ``m`` while
+instantiating a later variable, for instance ``params ~ ()``. Here we propose a
+very restricted use of explicit ``forall`` syntax to introduce type binders
+which can then be used to fill in the type arguments of a function which we'd
+like to quantify over, ::
+
+    {-# SPECIALISE forall m. foldsAndParser @m @() #-}
```

Comment:
Yes, I believe that should be fine. The only question is whether we want to allow the introduction of new constraints, I believe. That being said, I haven't thought terribly hard about what subtleties the `forall` syntax might introduce.